compiler/rustc_infer/src/infer/outlives/env.rs

```diff
@@ -50,22 +51,38 @@ pub struct OutlivesEnvironment<'tcx> {
     region_bound_pairs: RegionBoundPairs<'tcx>,
 }
 
+/// Builder of OutlivesEnvironment. Use this structure if you need to add more outlives
+/// bounds than `explicit_outlives_bounds(param_env)`.
+pub struct OutlivesEnvironmentBuilder<'tcx> {
+    pub param_env: ty::ParamEnv<'tcx>,
+    region_relation: TransitiveRelationBuilder<Region<'tcx>>,
+
+    region_bound_pairs: RegionBoundPairs<'tcx>,
+}
+
 /// "Region-bound pairs" tracks outlives relations that are known to
 /// be true, either because of explicit where-clauses like `T: 'a` or
 /// because of implied bounds.
 pub type RegionBoundPairs<'tcx> = Vec<(Region<'tcx>, GenericKind<'tcx>)>;
 
-impl<'a, 'tcx> OutlivesEnvironment<'tcx> {
-    pub fn new(param_env: ty::ParamEnv<'tcx>) -> Self {
-        let mut env = OutlivesEnvironment {
+impl<'tcx> OutlivesEnvironment<'tcx> {
+    /// Create a builder using `ParamEnv` and add explicit outlives bounds into it.
+    pub fn builder(param_env: ty::ParamEnv<'tcx>) -> OutlivesEnvironmentBuilder<'tcx> {
```

Comment:
This is only ever used to be immediately followed by `add_outlives_bounds`. Maybe make it private and expose a `with_implied_bounds` public constructor that just builds the `OutlivesEnvironment` in one go instead of

```rust
        let mut outlives_environment = OutlivesEnvironment::builder(param_env);
        outlives_environment.add_implied_bounds(infcx, x, y);
        let outlives_environment = outlives_environment.build();
```

being manually done like 5 times

Reply:
That's a good idea. I will make the change